File: 2/main.py
#!/usr/bin/python3

# 2727
# 56580

import sys

possible_ids = []
powers = []

for line in sys.stdin.readlines():
    line = line.strip()

    game, draws = line.split(':')
    game = int(game.split(' ')[1])
    draws = draws.split(';')

    possible = True
    count_map = {'red': 0, 'green': 0, 'blue': 0}

    for draw in draws:
        draw = draw.strip()
        takes = draw.split(',')

        for take in takes:
            cnt, color = take.strip().split(' ')
            cnt = int(cnt)

            if color == 'red' and cnt > 12:
                possible = False
            elif color == 'green' and cnt > 13:
                possible = False
            elif color == 'blue' and cnt > 14:
                possible = False

            count_map[color] = max(count_map[color], cnt)

    # print(f'Game {game} is possible {possible}')
    if possible:
        possible_ids.append(game)

    # print(f'Count map {count_map}')

    power = 1
    for cnt in count_map.values():
        power *= cnt
    powers.append(power)

print(sum(possible_ids))
print(sum(powers))

File: 2/main.hs
#!/usr/bin/env stack
{- stack script
   --resolver lts-21.22
   --package split
   --package text
-}

--  2727

import Data.Char
import Data.List.Split
import Data.Text (strip)

main :: IO ()
main = do
  input <- getContents
  let linesList = lines input
  let games = map parseLine linesList

  let possibleGames = filter isGamePossible games

  print $ sum $ map fst possibleGames

isGamePossible :: (Int, String) -> Bool
isGamePossible (gameId, game) = all isValidDraw draws
  where
    draws = splitOn ";" game

isValidDraw :: String -> Bool
isValidDraw draw = all isValidTake takes
  where
    takes = map parseTake (splitOn "," draw)

parseTake :: String -> (Int, String)
parseTake str = (cnt, color)
  where
    split = splitOn " " (trim str)
    cnt = read (head split) :: Int
    color = last split

trim :: String -> String
trim = f . f
  where
    f = reverse . dropWhile isSpace

isValidTake :: (Int, String) -> Bool
isValidTake (cnt, "red") = cnt <= 12
isValidTake (cnt, "green") = cnt <= 13
isValidTake (cnt, "blue") = cnt <= 14

parseLine :: String -> (Int, String)
parseLine line = (gameId, game)
  where
    split = splitOn ":" line
    gameId = parseGameId $ head split
    game = last split

parseGameId :: String -> Int
parseGameId str = read (splitOn " " str !! 1) :: Int
